Musig2: add new MuSig2 RPC methods to `signrpc`

guggero opened the pull request:
Depends on https://github.com/btcsuite/btcd/pull/1820.

Adds a new MuSig2 API for cooperatively producing Schnorr signatures that can be used either for Taproot internal keys or for `OP_CHECKSIG` signatures within a tap script tree.

Roasbeef commented on lnrpc/signrpc/signer.proto:
```diff
@@ -265,3 +301,125 @@ message SharedKeyResponse {
+    Indicates whether all nonces required to start the signing process are known
+    now.
+    */
+    bool have_all_nonces = 4;
```
Won't this always be false the first time a session is created?

guggero replied:
We can send a set of already known nonces. So for the last signer, we might already have all other nonces and therefore it can be true.

Roasbeef commented on lnrpc/signrpc/signer.proto:
```diff
@@ -265,3 +301,125 @@ message SharedKeyResponse {
+    of 66 bytes. Can be split into the two 33-byte points to get the individual
+    nonces.
+    */
+    bytes local_public_nonces = 3;
```
FWIW this is only actually needed for incremental verification, but maybe it's also useful on the caller side of things?

guggero replied:
We need to communicate the nonce to the other party, so we need to know it at some point, don't we? Or how would the nonce exchange work?

Roasbeef commented on lnrpc/signrpc/signer_server.go:
```diff
@@ -31,8 +36,33 @@ const (
+	// MuSig2SessionIDSize is the size of a MuSig2 session ID. It is the
+	// length of the combined public key and the two public nonces, all of
+	// them being compressed points.
+	MuSig2SessionIDSize = 99
```
Ah I see this sort of combines the notions of the session+context into one. If we wanted to get away with a smaller value there, then we can just hash these items and use a single 32 byte value here.

guggero replied:
Good idea with the hashing, added that.

Roasbeef reviewed (commented):
The final component! 

Completed an early initial pass focusing mainly on the protocol and client flow, which matches up pretty closely with what I had in mind.

lightninglabs-deploy commented:
@guggero, remember to re-request review from reviewers when ready

Roasbeef commented:
What needs to be done to take this out of the draft stage? Merging of the upstream btcd PR?

guggero commented:
> What needs to be done to take this out of the draft stage? Merging of the upstream btcd PR?

Mostly adding the MuSig2 methods to the signer interface so they can also be forwarded to the remote signer. And cleaning up some user input validation TODOs. So not a big diff, working on it today.

guggero commented:
Took the PR out of draft state. Everything works now, including remote signing!

While working on remote signing support for MuSig2 I noticed that the standalone Taproot functionality does not work in remote mode. Turns out we need https://github.com/lightningnetwork/lnd/issues/6268 to enable that.
I'm going to work on that in a separate PR, commented out the failing remote mode itests for now.

bhandras reviewed (approved):
The only thing remaining is to make the linter happy and fix the mobile build.

Other than those it's a super clean PR,  really enjoyed reviewing! Awesome addition to LND :rocket:  Great work @guggero :1st_place_medal:

lightninglabs-deploy commented:
@roasbeef: review reminder

guggero commented:
I added a convenience RPC called `MuSig2CombineKeys` to make it possible to just get the combined key and Taproot public key without creating a session.

I also went ahead and added a short doc explaining the current MuSig2 implementation and examples.

guggero commented:
I updated to the latest version of https://github.com/btcsuite/btcd/pull/1820 and added an itest for using a MuSig2 combined key in a tap leaf with a simple `OP_CHECKSIG` script.

I also found a small bug in the helper function `TapscriptPartialReveal` that is used for the fee calculation. A proof for a tree with just a single leaf can be empty, so the inclusion proof can be variable size.

Roasbeef commented:
The btcd PR has been merged, so this can be rebased!

Roasbeef commented:
> A proof for a tree with just a single leaf can be empty, so the inclusion proof can be variable size.

Yes in this case, the verifier just takes the leaf, computes one level of tapLeaf hashing and uses that directly. You can see this interaction here if you pretend the loop below doesn't actually exist: https://github.com/btcsuite/btcd/blob/master/txscript/taproot.go#L155-L177

Roasbeef commented on lntest/itest/lnd_remote_signer_test.go:
```diff
@@ -114,6 +114,29 @@ func testRemoteSigner(net *lntest.NetworkHarness, t *harnessTest) {
+			)
+			defer cancel()
+
+			// TODO(guggero): Fix remote taproot signing by adding
```
👍

Roasbeef commented on lntest/itest/lnd_taproot_test.go:
```diff
@@ -337,6 +341,524 @@ func testTaprootSignOutputRawKeySpendRootHash(ctxt context.Context,
+
+	// Serialize, weigh and publish the TX now, then make sure the
+	// coins are sent and confirmed to the final sweep destination address.
+	publishTxAndConfirmSweep(
```
🤘

Roasbeef commented on docs/musig2.md:
```diff
@@ -0,0 +1,117 @@
+# MuSig2
+
+With `lnd v0.15.0-beta` a new, experimental MuSig2 API was added to the
```
Great documentation 👍

Roasbeef reviewed (approved):
LGTM 🌂

Roasbeef commented:
The new `btcec` tag: https://github.com/btcsuite/btcd/releases/tag/btcec%2Fv2.2.0

guggero commented on go.mod:
```diff
@@ -50,22 +48,116 @@ require (
+require (
+	github.com/NebulousLabs/fastrand v0.0.0-20181203155948-6fb6489aac4e // indirect
+	github.com/aead/chacha20 v0.0.0-20180709150244-8b13a72661da // indirect
+	github.com/aead/siphash v1.0.1 // indirect
```
No, unfortunately not. Go 1.17/1.18 just puts all the indirect dependencies in their own block.

guggero commented on input/musig2.go:
```diff
@@ -0,0 +1,228 @@
+
+// MuSig2SessionID is a type for a session ID that is just a hash of the MuSig2
+// combined key and the local public nonces.
+type MuSig2SessionID = [sha256.Size]byte
```
Ah, good point. I thought the alias would make it easier to convert between the type and the hash returned by the `sha256` package. But I don't think that's really required.

About your second comment: I'll address that in a follow-up PR. Would like to get this PR in as is to unblock some work down the dependency tree.

guggero commented on input/musig2.go:
```diff
@@ -0,0 +1,228 @@
+	// already known, they can be submitted as well to reduce the number of
+	// method calls necessary later on.
+	MuSig2CreateSession(keychain.KeyLocator, []*btcec.PublicKey,
+		*MuSig2Tweaks, [][musig2.PubNonceSize]byte) (*MuSig2SessionInfo,
```
No, we don't make the distinction because with the previous design of the API we needed the session to get our local nonce so we could give it to the other parties. 

The list of other party public nonces is not optional as in dynamic argument list. But you can pass in `nil` if you're the first to create a session and don't know any nonces of other participants yet.

I'll see how this looks once I've incorporated the recent API changes in the context/session. I hope I can add that functionality with minimal changes to the RPC API surface.
